[MLIR][Linalg] Introduce transpose/broadcast semantic to linalg.batch…

shahidact opened the pull request:
…_reduce_matmul.

This patch exposes broadcast and transpose semantics on 'batch_reduce_matmul'. This is the last one in continuation of other two variant of matmul ops.

The broadcast and transpose semantic are as follows:

Broadcast and Transpose semantics can be appiled by specifying the explicit attribute 'indexing_maps' as shown below. This is a list attribute, so must include maps for all arguments if specified.

    Example Transpose:
    ```
    linalg.batch_reduce_matmul indexing_maps = [
       affine_map<(d0, d1, d2, d3) -> (d0, d3, d1)>, // transpose
       affine_map<(d0, d1, d2, d3) -> (d0, d3, d2)>,
       affine_map<(d0, d1, d2, d3) -> (d1, d2)>
       ]
          ins(%arg0, %arg1 : memref<2x5x3xf32>,memref<2x5x7xf32>)
          outs(%arg2: memref<3x7xf32>)
    ```

    Example Broadcast:
    ```
    linalg.batch_reduce_matmul indexing_maps = [
       affine_map<(d0, d1, d2, d3) -> (d3)>,         // broadcast
       affine_map<(d0, d1, d2, d3) -> (d0, d3, d2)>,
       affine_map<(d0, d1, d2, d3) -> (d1, d2)>
       ]
          ins(%arg0, %arg1 : memref<5xf32>, memref<2x5x7xf32>)
          outs(%arg2: memref<3x7xf32>)
    ```

    Example Broadcast and Transpose:
    ```
    linalg.batch_reduce_matmul indexing_maps = [
       affine_map<(d0, d1, d2, d3) -> (d1, d3)>,     // broadcast
       affine_map<(d0, d1, d2, d3) -> (d0, d2, d3)>, // transpose
       affine_map<(d0, d1, d2, d3) -> (d1, d2)>
       ]
          ins(%arg0, %arg1 : memref<3x5xf32>, memref<2x7x5xf32>)
          outs(%arg2: memref<3x7xf32>)
    ```

RFCs and related PR:
https://discourse.llvm.org/t/rfc-linalg-opdsl-constant-list-attribute-definition/80149
https://discourse.llvm.org/t/rfc-op-explosion-in-linalg/82863
https://discourse.llvm.org/t/rfc-mlir-linalg-operation-tree/83586
https://github.com/llvm/llvm-project/pull/115319
https://github.com/llvm/llvm-project/pull/122275

shahidact committed:
[MLIR][Linalg] Introduce transpose/broadcast semantic to linalg.batch_reduce_matmul.

This patch exposes broadcast and transpose semantics on
'batch_reduce_matmul'. This is the last one in continuation of
other two variant of matmul ops.

The broadcast and transpose semantic are as follows:

Broadcast and Transpose semantics can be appiled by specifying the
explicit attribute 'indexing_maps' as shown below. This is a list
attribute, so must include maps for all arguments if specified.

    Example Transpose:
    ```
    linalg.batch_reduce_matmul indexing_maps = [
       affine_map<(d0, d1, d2, d3) -> (d0, d3, d1)>, // transpose
       affine_map<(d0, d1, d2, d3) -> (d0, d3, d2)>,
       affine_map<(d0, d1, d2, d3) -> (d1, d2)>
       ]
          ins(%arg0, %arg1 : memref<2x5x3xf32>,memref<2x5x7xf32>)
          outs(%arg2: memref<3x7xf32>)
    ```

    Example Broadcast:
    ```
    linalg.batch_reduce_matmul indexing_maps = [
       affine_map<(d0, d1, d2, d3) -> (d3)>,         // broadcast
       affine_map<(d0, d1, d2, d3) -> (d0, d3, d2)>,
       affine_map<(d0, d1, d2, d3) -> (d1, d2)>
       ]
          ins(%arg0, %arg1 : memref<5xf32>, memref<2x5x7xf32>)
          outs(%arg2: memref<3x7xf32>)
    ```

    Example Broadcast and Transpose:
    ```
    linalg.batch_reduce_matmul indexing_maps = [
       affine_map<(d0, d1, d2, d3) -> (d1, d3)>,     // broadcast
       affine_map<(d0, d1, d2, d3) -> (d0, d2, d3)>, // transpose
       affine_map<(d0, d1, d2, d3) -> (d1, d2)>
       ]
          ins(%arg0, %arg1 : memref<3x5xf32>, memref<2x7x5xf32>)
          outs(%arg2: memref<3x7xf32>)
    ```

RFCs and related PR:
https://discourse.llvm.org/t/rfc-linalg-opdsl-constant-list-attribute-definition/80149
https://discourse.llvm.org/t/rfc-op-explosion-in-linalg/82863
https://discourse.llvm.org/t/rfc-mlir-linalg-operation-tree/83586
https://github.com/llvm/llvm-project/pull/115319
https://github.com/llvm/llvm-project/pull/122275

banach-space commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:
```diff
@@ -1054,6 +1054,137 @@ def BatchMatmulOp : LinalgStructuredBase_Op<"batch_matmul", !listconcat([AttrSiz
+                          LinalgContractionOpInterface]> {
+    
+  let summary = [{Performs a batch-reduce matrix multiplication of two 3D inputs.
+The partial multiplication results are reduced into a 2D output.}];
```
Please fix indentation.

banach-space commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:
````diff
@@ -1054,6 +1054,137 @@ def BatchMatmulOp : LinalgStructuredBase_Op<"batch_matmul", !listconcat([AttrSiz
+
+    Example Transpose:
+    ```
+    linalg.batch_reduce_matmul indexing_maps = [
+                   affine_map<(d0, d1, d2, d3) -> (d0, d3, d1)>, // transpose
+                   affine_map<(d0, d1, d2, d3) -> (d0, d3, d2)>,
+                   affine_map<(d0, d1, d2, d3) -> (d1, d2)>
+                   ]
+                   ins(%arg0, %arg1 : memref<2x5x3xf32>,memref<2x5x7xf32>)
+                   outs(%arg2: memref<3x7xf32>)
````
This indentation is unnecessarily deep. Could we follow the "golden" standard of `linalg.generic`, please?
* https://mlir.llvm.org/docs/Dialects/Linalg/#linalggeneric-linalggenericop

`linalg.contract` is also more consistently formatted (and looks very similar to `linalg.batch_reduce_matmul`:
* https://mlir.llvm.org/docs/Dialects/Linalg/#linalgcontract-linalgcontractop

So, this should look like:
```mlir
    linalg.batch_reduce_matmul
      indexing_maps = [ affine_map<(d0, d1, d2, d3) -> (d0, d3, d1)>, // transpose
                        affine_map<(d0, d1, d2, d3) -> (d0, d3, d2)>,
                        affine_map<(d0, d1, d2, d3) -> (d1, d2)>]
      ins(%arg0, %arg1 : memref<2x5x3xf32>,memref<2x5x7xf32>)
      outs(%arg2: memref<3x7xf32>)
```

Very importantly, this file should use consistent formatting. I appreciate that that's currently not the case:
* https://github.com/llvm/llvm-project/blob/d928a671b84afb9c2ad64353694537a198f04651/mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td#L857-L864 vs
* https://github.com/llvm/llvm-project/blob/d928a671b84afb9c2ad64353694537a198f04651/mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td#L714-L721

My ask is that we begind converging towards uniform style starting with this PR. To me, `linalg.generic` and `linalg.contract` are good examples to follow.

shahidact replied:
Done.

banach-space commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:
```diff
@@ -1054,6 +1054,137 @@ def BatchMatmulOp : LinalgStructuredBase_Op<"batch_matmul", !listconcat([AttrSiz
+    Numeric casting is performed on the operands to the inner multiply, promoting
+    them to the same data type as the accumulator/output.
+
+    Broadcast and Transpose semantics can be appiled by specifying the explicit attribute
+    'indexing_maps' as shown below. This is a list attribute, so must include maps for all
+    arguments if specified.
```
I presume the broadcast and transpose semantics are going to be identical to `linalg.contract`? If yes, can we re-use this comment: https://github.com/llvm/llvm-project/blob/d928a671b84afb9c2ad64353694537a198f04651/mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td#L851-L855

shahidact replied:
I would prefer to do these semantic related comment update in later NFC patch than the current one.

banach-space commented on mlir/test/Dialect/Linalg/generalize-named-ops.mlir:
```diff
@@ -1024,6 +1024,36 @@ func.func @batch_matmul(%arg0: tensor<2x3x5xf32>, %arg1: tensor<2x5x7xf32>, %arg
+// CHECK:             %[[VAL_8:.*]] = arith.addf %[[VAL_6]], %[[VAL_7]] : f32
+// CHECK:             linalg.yield %[[VAL_8]] : f32
+// CHECK:           } -> tensor<3x7xf32>
+// CHECK:           return %[[VAL_3]] : tensor<3x7xf32>
+// CHECK:         }
```
Please only check what's necessary: https://mlir.llvm.org/getting_started/TestingGuide/#filecheck-best-practices

shahidact replied:
Done.

banach-space commented on mlir/test/Dialect/Linalg/generalize-named-ops.mlir:
```diff
@@ -1024,6 +1024,36 @@ func.func @batch_matmul(%arg0: tensor<2x3x5xf32>, %arg1: tensor<2x5x7xf32>, %arg
+// CHECK: #[[$ATTR_2:.+]] = affine_map<(d0, d1, d2, d3) -> (d1, d2)>
+
+// CHECK-LABEL:   func.func @batch_reduce_matmul(
+// CHECK-SAME:                                   %[[VAL_0:[0-9]+|[a-zA-Z$._-][a-zA-Z0-9$._-]*]]: tensor<2x3x5xf32>,
+// CHECK-SAME:                                   %[[VAL_1:[0-9]+|[a-zA-Z$._-][a-zA-Z0-9$._-]*]]: tensor<2x5x7xf32>,
+// CHECK-SAME:                                   %[[VAL_2:[0-9]+|[a-zA-Z$._-][a-zA-Z0-9$._-]*]]: tensor<3x7xf32>) -> tensor<3x7xf32> {
```
Please avoid unnecessary deep indentation

shahidact replied:
Done.

banach-space commented on mlir/test/Dialect/Linalg/generalize-named-ops.mlir:
```diff
@@ -1024,6 +1024,36 @@ func.func @batch_matmul(%arg0: tensor<2x3x5xf32>, %arg1: tensor<2x5x7xf32>, %arg
+// CHECK-SAME:                                   %[[VAL_0:[0-9]+|[a-zA-Z$._-][a-zA-Z0-9$._-]*]]: tensor<2x3x5xf32>,
+// CHECK-SAME:                                   %[[VAL_1:[0-9]+|[a-zA-Z$._-][a-zA-Z0-9$._-]*]]: tensor<2x5x7xf32>,
+// CHECK-SAME:                                   %[[VAL_2:[0-9]+|[a-zA-Z$._-][a-zA-Z0-9$._-]*]]: tensor<3x7xf32>) -> tensor<3x7xf32> {
+// CHECK:           %[[VAL_3:.*]] = linalg.generic {indexing_maps = [#[[$ATTR_0]], #[[$ATTR_1]], #[[$ATTR_2]]], iterator_types = ["reduction", "parallel", "parallel", "reduction"]} ins(%[[VAL_0]], %[[VAL_1]] : tensor<2x3x5xf32>, tensor<2x5x7xf32>) outs(%[[VAL_2]] : tensor<3x7xf32>) {
```
[nit] Split across multiple lines

shahidact replied:
Done.

banach-space commented on mlir/test/Dialect/Linalg/generalize-named-ops.mlir:
Please use meaningful LIT and MLIR variable names: https://mlir.llvm.org/getting_started/TestingGuide/#test-formatting-best-practices

shahidact replied:
Done.

banach-space commented on mlir/test/Dialect/Linalg/invalid.mlir:
```diff
@@ -1462,6 +1462,208 @@ func.func @invalid_C_map_result_dim_batch_matmul(%arg0: memref<?x?x?xf32>, %arg1
 
+// -----
+
+func.func @indexing_map_size_mismatch_batch_reduce_matmul(%arg0: memref<?x?x?xf32>,
```
The size of which indexing map is wrong? Or is it the size of the attribute? If it's the latter, then `@indexing_map_size_mismatch_` -> `@missing_indexing_map_`?

shahidact replied:
Done.

banach-space commented on mlir/test/Dialect/Linalg/invalid.mlir:
```diff
@@ -1462,6 +1462,208 @@ func.func @invalid_C_map_result_dim_batch_matmul(%arg0: memref<?x?x?xf32>, %arg1
+
+func.func @missing_indexing_map_batch_reduce_matmul(%arg0: memref<?x?x?xf32>, %arg1: memref<?x?x?xf32>, %arg2: memref<?x?xf32>) {
+  // expected-error @+1 {{expected attribute value}}
+  linalg.batch_reduce_matmul indexing_maps = [
+                       ,
+                       affine_map<(d0, d1, d2, d3) -> (d0, d3, d2)>,
+                       affine_map<(d0, d1, d2, d3) -> (d1, d2)>
+                      ]
+                      ins(%arg0, %arg1 : memref<?x?x?xf32>, memref<?x?x?xf32>)
+                      outs(%arg2 :memref<?x?xf32>)
```
Here and other places, please improve indentation.

shahidact replied:
Done.

banach-space commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:
````diff
@@ -1054,6 +1054,137 @@ def BatchMatmulOp : LinalgStructuredBase_Op<"batch_matmul", !listconcat([AttrSiz
+    Example Transpose:
+    ```
+    linalg.batch_reduce_matmul indexing_maps = [
+                   affine_map<(d0, d1, d2, d3) -> (d0, d3, d1)>, // transpose
````
Could we use a more self-descriptive naming convention? I suggest: https://github.com/llvm/llvm-project/blob/d928a671b84afb9c2ad64353694537a198f04651/mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td#L844-L846

shahidact replied:
Done.

banach-space reviewed (commented):
Thanks for the contribution and apologies for the delay reviewing!

From what I can tell, the semantics match what we have already introduced for other Ops (e.g. `linalg.matmul` + `linalg.contract`). 

Since it's a Friday, I'm starting with the easier part - formatting 😅  I will go over C++ later.

MaheshRavishankar commented:
Thanks for the ping. I skimmed through the change and I think it looks fine to me. I am not rely in a position to do a deep review soon-ish, but it seems consistent with the `linalg.contract`.

shahidact commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:
```diff
@@ -1054,6 +1054,137 @@ def BatchMatmulOp : LinalgStructuredBase_Op<"batch_matmul", !listconcat([AttrSiz
+  let summary = [{Performs a batch-reduce matrix multiplication of two 3D inputs.
+The partial multiplication results are reduced into a 2D output.}];
+  let description = [{
+    Numeric casting is performed on the operands to the inner multiply, promoting
+    them to the same data type as the accumulator/output.
```
As the scope is to keep the existing cast semantic, signed casting is performed.

shahidact committed:
1. Improved indentation in op definition description and test cases.

nicolasvasilache commented:
Thanks for your PR, in its current form I believe it will break the python side of things though.
Please update and add a test on the python side if you remove the opdsl / yaml source of truth.

shahidact committed:
-Added python test for linalg.batch_reduce_matmul

shahidact commented:
> Thanks for your PR, in its current form I believe it will break the python side of things though. Please update and add a test on the python side if you remove the opdsl / yaml source of truth.

Thanks for the review, updated accordingly.

makslevental commented on mlir/test/python/dialects/linalg/ops.py:
```diff
@@ -568,6 +568,107 @@ def batch_matmul_op(A, Amem, B, Bmem, Btransposed, Btransposedmem, C, Cmem):
+                    result_tensors=(C.type,),
+                    inputs=(A, Btransposed),
+                    outputs=(C,),
+                    indexing_maps=[a_map, b_transposed_map, c_map],
```
I'm just gonna ask the same question I've been asking over and over on these PRs: why is the API changing such that we need to provide indexing maps to a named matmul operation? 

I also don't understand why no one seems to recognize this as a problem - isn't the entire point of named ops that they don't require the user to supply the indexing map? If in the near future all named linalg ops will also require users to provide indexing maps then are we effectively eliminating all named ops?

nicolasvasilache replied:
@makslevental is your concern about the direction in general or about the lack of default "no-indexing map == canonical identity" APIs?

If it is the latter, this is an easy fix.
It if is the former, my perspective is a bunch of folks have an itch to displace matching code towards C++ verifier.
I am fine with that as it does indeed make IR quite more readable with fewer generics, even if the improvement in code logic is very marginal.

It would be even more readable if the affine_map actually pretty printed with the names batch, m, n, k to easily see what is what. I'd welcome a followup improvement to the pretty-printer (like op result can have asm overrides), @rengolin @ftynse what say you?

While we're at pretty-printer wishlists, if would be nice to pretty print the op in a formatted fashion. I'd love to have my named linalg ops be pretty-printed exactly like this:
```
func.func @missing_indexing_map_batch_reduce_matmul(%arg0: memref<?x?x?xf32>, %arg1: memref<?x?x?xf32>, %arg2: memref<?x?xf32>) {
      linalg.batch_reduce_matmul
          indexing_maps = [affine_map<(batch, m, n, k) -> (batch, m, k)>,
                           affine_map<(batch, m, n, k) -> (batch, n, k)>]
          ins(%arg0, %arg1 : memref<?x?x?xf32>, memref<?x?x?xf32>)
          outs(%arg2: memref<?x?xf32>)
      return
 }
```

makslevental replied:
I am indeed talking about the latter and I'm glad to hear it's an easy fix because that's exactly what I would expect (and what I did expect in the previous PR).

makslevental replied:
> I'd welcome others that have more familiarity with the code (and Python) to make the necessary adjustments, maybe after this PR.

Ok but these are breaking changes are they not? Just like the last one? Like I couldn't just change some python that affected the cpp codebase and then ask people to fix the cpp just because I'm not familiar with that code right?

makslevental replied:
Okay after actually reading a little I see that there is `BatchReduceMatmulOp::getDefaultIndexingMaps` and that all that it would take would be to plumb that through to C/Python. Okay I can sign up for doing that (this is me signing up) after this PR lands.

makslevental replied:
Okay after actually reading a little I see that there is `BatchReduceMatmulOp::getDefaultIndexingMaps` and that all that it would take would be to plumb that through to C/Python. Okay I can sign up for doing that (this is me signing up) after this PR lands.

shahidact replied:
@banach-space @nicolasvasilache IMO, addressed all comments. if anything left, pls let me know.

rengolin commented:
Ping

banach-space commented on mlir/test/Dialect/Linalg/invalid.mlir:
```diff
@@ -1484,6 +1484,201 @@ func.func @invalid_C_map_result_dim_batch_matmul(%arg0: memref<?x?x?xf32>, %arg1
 
+// -----
+
+func.func @missing_indexing_map_batch_reduce_matmul(%arg0: memref<?x?x?xf32>,
```
Rather than repeating `batch_reduce_matmul` in every test function name, could you add a block comment? For example:
```mlir
//===----------------------------------------------------------------------===//
// linalg.batch_reduce_matmul
//===----------------------------------------------------------------------===//
```

shahidact replied:
Done.

banach-space commented on mlir/test/Dialect/Linalg/invalid.mlir:
```diff
@@ -1484,6 +1484,201 @@ func.func @invalid_C_map_result_dim_batch_matmul(%arg0: memref<?x?x?xf32>, %arg1
+
+// -----
+
+func.func @indexing_map_size_one_batch_reduce_matmul(%arg0: memref<?x?x?xf32>,
+     %arg1: memref<?x?x?xf32>, %arg2: memref<?x?xf32>) {
+     // expected-error @+1 {{Indexing_map attribute must have 3 affine maps}}
+     linalg.batch_reduce_matmul
+         indexing_maps = [affine_map<(batch, m, n, k) -> (batch, m, k)>]
+         ins(%arg0, %arg1 : memref<?x?x?xf32>, memref<?x?x?xf32>)
+         outs(%arg2: memref<?x?xf32>)
+     return
+
+}
```
IIUC, this test is identical to the one above, but 2 maps are missing rather than one? If yes:
* `@missing_indexing_map_size_two_batch_reduce_matmul`

Or something similar to maintain consistency.

shahidact replied:
Done.

banach-space commented on mlir/test/Dialect/Linalg/invalid.mlir:
```diff
@@ -1484,6 +1484,201 @@ func.func @invalid_C_map_result_dim_batch_matmul(%arg0: memref<?x?x?xf32>, %arg1
+
+// -----
+
+func.func @invalid_dim_expr_batch_reduce_matmul_a(%arg0: memref<?x?x?xf32>, %arg1: memref<?x?x?xf32>, %arg2: memref<?x?xf32>) {
```
[nit] `a` in `@missing_indexing_map_batch_reduce_matmul_a` refers to matrix `A` (please use caps for matrices), but the argument name is `%arg0`. This is confusing, could you unify? Thanks!

shahidact replied:
Done.

banach-space commented on mlir/test/Dialect/Linalg/invalid.mlir:
```diff
@@ -1484,6 +1484,201 @@ func.func @invalid_C_map_result_dim_batch_matmul(%arg0: memref<?x?x?xf32>, %arg1
+// -----
+
+func.func @invalid_bcast_batch_reduce_matmul_a(%arg0: memref<?xf32>, %arg1: memref<?x?x?xf32>, %arg2: memref<?x?xf32>) {
+  // expected-error @+1 {{'linalg.batch_reduce_matmul' op Invalid broadcast requested}}
```
Either include Op name everywhere or skip it everywhere.
```suggestion
  // expected-error @+1 {{Invalid broadcast requested}}
```

banach-space commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:
````diff
@@ -1065,6 +1060,134 @@ def BatchMatmulOp : LinalgStructuredBase_Op<"batch_matmul", !listconcat([AttrSiz
+    Example Transpose:
+    ```
+    linalg.batch_reduce_matmul
+        indexing_maps = [affine_map<(d0, d1, d2, d3) -> (d0, d3, d1)>, // transpose
+                         affine_map<(d0, d1, d2, d3) -> (d0, d3, d2)>,
+                         affine_map<(d0, d1, d2, d3) -> (d1, d2)>]
````
[nit] Would you mind using `batch`, `m`, `n` and `k` for dim expressions like you did in tests? That's so much more readable. I'd say that this part is more important than tests. Thanks!

shahidact replied:
Done.

banach-space commented on mlir/test/Dialect/Linalg/invalid.mlir:
```diff
@@ -1484,6 +1484,201 @@ func.func @invalid_C_map_result_dim_batch_matmul(%arg0: memref<?x?x?xf32>, %arg1
+func.func @invalid_multi_dim_bcast_expr_batch_reduce_matmul_a(%arg0: memref<?x?xf32>, %arg1: memref<?x?x?xf32>, %arg2: memref<?x?xf32>) {
+  // expected-error @+1 {{'linalg.batch_reduce_matmul' op Invalid broadcast requested}}
+  linalg.batch_reduce_matmul
+      indexing_maps = [affine_map<(batch, m, n, k) -> (batch, k)>,
```
Presumably this is the incorrect broadcast? Why is it incorrect?

shahidact replied:
Done.

shahidact replied:
> My question stands :) What makes this an invalid broadcast?

`affine_map<(batch, m, n, k) -> (k, batch)`, This is invalid as it would not represent a valid matrix multiplication mathematically, as the dimensions that should be multiplied (k in A and k in B) are in positions that don't align properly for the operation.

banach-space replied:
Apologies, I forgot what the definition of "valid" was and hence my confusion (initially I wasn't sure what this was checking). IIUC, that's specified here: https://github.com/llvm/llvm-project/blob/082598a64d1f3d583c7ae979681df90f9f9c0f63/mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp#L3943-L3959

You are merely checking against that definition, so all good. That said, the documentation of `:isValidLhsRhsBroadcastMap` is way too sparse though and we should update it. I can do it, but can you remind me why we allow this:
```mlir
indexing_maps = [affine_map<(batch, m, n, k) -> (batch, k)>,
```
but reject this:
```mlir
indexing_maps = [affine_map<(batch, m, n, k) -> (batch, m)>,
```
?

shahidact replied:
`indexing_maps = [affine_map<(batch, m, n, k) -> (batch, m)>`, This is not mathematically valid for any matmul ops as common k dimension is missing which is essential to represent matmul ops.

shahidact replied:
Apologies for not being clear enough. Matrix multiplication is defined as the sum of products along a shared dimension. The shared dimension is K which must be same for both A and B matrices while M and N can differ. For example, if M not equal to N and we broadcast M to K for A matrix and N to K for B matrix, K no longer remains common/shared, consequently violating the `matmul` definition.

banach-space replied:
>  For example, if M not equal to N and we broadcast M to K for A matrix and N to K for B matrix, K no longer remains common/shared, consequently violating the matmul definition.

I don't quite follow this - this sounds like we could end up with ... two different `K`s?

In general, there are two possibilities:
* `A = tensor<M x i32>` and `B = tensor<K x N x i32>` - what's the problem with broadcasting `A` to `[M x K]` where `K` is taken from `B`?
* `A = tensor<M x K x i32>` and `B = tensor<N x i32>` - what's the problem with broadcasting `B` to `[K x N]` where `K` is taken from `A`?

There's a 3rd option that indeed does not make sense (there's nowhere to obtain `K` from):
* `A = tensor<M x i32>` and `B = tensor<N x i32>`.

Is this consistent with your understanding? If yes, what ma I missing here?

shahidact replied:
> * `A = tensor<M x i32>` and `B = tensor<K x N x i32>` - what's the problem with broadcasting `A` to `[M x K]` where `K` is taken from `B`?
> * `A = tensor<M x K x i32>` and `B = tensor<N x i32>` - what's the problem with broadcasting `B` to `[K x N]` where `K` is taken from `A`?

IMO, no problem in this but we are not handling this as of now.

shahidact replied:
> Is this consistent with your understanding? If yes, what ma I missing here?

Yes, I was missing your perspective.

banach-space replied:
> IMO, no problem in this but we are not handling this as of now.

Great, could you add TODOs in relevant places? Otherwise this restriction feels ad-hoc.

shahidact replied:
Done.

banach-space commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -5338,6 +5383,166 @@ struct FoldTensorCastUnPackOp : public OpRewritePattern<UnPackOp> {
+    p << "]";
+  }
+
+  SmallVector<StringRef, 3> elidedAttrs = {
+      "operandSegmentSizes", "linalg.memoized_indexing_maps", "indexing_maps"};
+  ::printNamedStructuredOp(p, getOperation(), getInputs(), getOutputs(),
+                           elidedAttrs);
```
Hm, I don't see this appearing in tests. What is it? :)

shahidact replied:
By default, printer exclude these attributes from printing. The explicit `indexing_maps` are printed if it is not same as default `indexing_maps`.

shahidact replied:
> `memoized_indexing_maps`

IIUC this is to cache the indexing_map array attribute. It gets printed on diagnostics. @nicolasvasilache could you pls enlighten us.

banach-space reviewed (commented):
Sorry for the delay — returning to this after a short break.

Most of my comments are related to test hygiene; they’re minor, but still worth addressing.

While reviewing invalid.mlir, I got a bit confused by the "invalid broadcast" case — hopefully you can clarify that (I probably just need to refresh my memory). On a related note: since the verification logic is shared, shouldn’t the invalid tests for `linalg.batch_matmul` and `linalg.reduce_batch_matmul` be nearly identical?

shahidact commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:
```diff
@@ -1065,6 +1060,134 @@ def BatchMatmulOp : LinalgStructuredBase_Op<"batch_matmul", !listconcat([AttrSiz
+      static void regionBuilder(ImplicitLocOpBuilder &b,
+                                Block &block, ArrayRef<NamedAttribute> attrs);
+
+      /// Returns a list of AffineMap with the typical batch_reducematmul indexing charactristic.
```
With `typical`, I meant the default `indexing_maps `for this op. `batch_reducematmul` is a typo. I will correct it.

banach-space replied:
> I meant the default

I suggest using "default" to disambiguate :)

shahidact replied:
Done.

shahidact committed:
-Fixed invalid broadcast test.
-Added consistent variable and function naming in test cases.
-Improved ops indexing_maps description.

shahidact commented on mlir/test/Dialect/Linalg/generalize-named-ops.mlir:
```diff
@@ -1024,6 +1024,34 @@ func.func @batch_matmul(%arg0: tensor<2x3x5xf32>, %arg1: tensor<2x5x7xf32>, %arg
+// CHECK:           arith.addf
+// CHECK:           linalg.yield
+
+func.func @batch_reduce_matmul(%arg0: tensor<2x3x5xf32>, %arg1: tensor<2x5x7xf32>, %arg2: tensor<3x7xf32>) -> tensor<3x7xf32> {
```
Done.

shahidact commented on mlir/test/Dialect/Linalg/named-ops.mlir:
```diff
@@ -1637,6 +1637,171 @@ func.func @batch_matmul_bcast_A_transpose_B(%arg0: memref<3x5xf32>, %arg1: memre
+// CHECK-LABEL:   func.func @batch_reduce_matmul_bcast_k_to_fill_missing_dims_A(
+// CHECK-SAME:                                                                  %[[VAL_0:[0-9]+|[a-zA-Z$._-][a-zA-Z0-9$._-]*]]: memref<5xf32>,
+// CHECK-SAME:                                                                  %[[VAL_1:[0-9]+|[a-zA-Z$._-][a-zA-Z0-9$._-]*]]: memref<2x5x7xf32>,
+// CHECK-SAME:                                                                  %[[VAL_2:[0-9]+|[a-zA-Z$._-][a-zA-Z0-9$._-]*]]: memref<3x7xf32>) {
```
Done.

shahidact commented:
Hi @banach-space IMO, I addressed all the comment.

banach-space reviewed (approved):
Overall LGTM. I would like my two questions to be addressed before merging, but they are non-blocking, so I'm approving as is.

Please allow at least a few more days before landing - it would be good to hear from other reviewers as well.

Thanks for the effort - great to see all these improvements to Linalg!

shahidact committed:
-Formats example IR for the linalg.*matmul ops for syntax highlighting.

shahidact commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:
````diff
@@ -1065,6 +1060,134 @@ def BatchMatmulOp : LinalgStructuredBase_Op<"batch_matmul", !listconcat([AttrSiz
+    arguments if specified.
+
+    Example Transpose:
+    ```
````
Done.

shahidact commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:
````diff
@@ -1065,6 +1060,134 @@ def BatchMatmulOp : LinalgStructuredBase_Op<"batch_matmul", !listconcat([AttrSiz
+    ```
+
+    Example Broadcast:
+    ```
````
Done.

shahidact commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:
````diff
@@ -1065,6 +1060,134 @@ def BatchMatmulOp : LinalgStructuredBase_Op<"batch_matmul", !listconcat([AttrSiz
+    ```
+
+    Example Broadcast and Transpose:
+    ```
````
Done.

javedabsar1 commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:
```diff
@@ -690,34 +690,32 @@ def MatmulOp : LinalgStructuredBase_Op<"matmul", [
-                   ]
-                   ins(%arg0, %arg1 : memref<5x3xf32>,memref<5x7xf32>)
-                   outs(%arg2: memref<3x7xf32>)
+    linalg.matmul
```
Would it be better to upfront mention
`each map can be a projected-permutation (simple transpose or broadcast, or mix of two) as long as the dimensions are application of the map match the semantics of batch_matmul`

shahidact replied:
> Would it be better to upfront mention `each map can be a projected-permutation (simple transpose or broadcast, or mix of two) as long as the dimensions are application of the map match the semantics of batch_matmul`

Do you mean to update the description as below?

`Broadcast and Transpose semantics can be applied by specifying the explicit projected permutation map using attribute 'indexing_maps' as shown below.`

banach-space replied:
> > Does `linalg.batch_reduce_matmul` differ from `linalg.contract`?
> 
> No.

Then why not re-use the definition from `linalg.contract` and begin standarisin things?

> Moreover, I would not prefer to change the documentation of matmul variants in this PR.

I've not noticed anyone proposing this.

shahidact replied:
> Then why not re-use the definition from `linalg.contract` and begin standarisin things?

Yes we can, however linalg.contract is newer addition relatively.

shahidact replied:
> If this is just about copying the docs, then we can just have the same across. Agree to not change other docs, but if the `contract` is better, we can reuse it here, too.

Yes, we will revisit the documentation in future.

javedabsar1 commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:
```diff
@@ -1065,6 +1060,134 @@ def BatchMatmulOp : LinalgStructuredBase_Op<"batch_matmul", !listconcat([AttrSiz
+                          AttrSizedOperandSegments,
+                          LinalgContractionOpInterface]> {
+    
+  let summary = [{Performs a batch-reduce matrix multiplication of two 3D inputs.
```
`3D` inputs is incongruent with broadcast possibility that the map provides.

javedabsar1 commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:
```diff
@@ -1065,6 +1060,134 @@ def BatchMatmulOp : LinalgStructuredBase_Op<"batch_matmul", !listconcat([AttrSiz
+
+    Broadcast and Transpose semantics can be appiled by specifying the explicit attribute
+    'indexing_maps' as shown below. This is a list attribute, so must include maps for all
+    arguments if specified.
```
maybe somethin like 
s/if specified/if any one needs non-default (specified)/

shahidact replied:
IMO, this already conveys the intent.

javedabsar1 commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:
```diff
@@ -1065,6 +1060,134 @@ def BatchMatmulOp : LinalgStructuredBase_Op<"batch_matmul", !listconcat([AttrSiz
+  let summary = [{Performs a batch-reduce matrix multiplication of two 3D inputs.
+    The partial multiplication results are reduced into a 2D output.}];
+  let description = [{
+    Numeric casting is performed on the operands to the inner multiply, promoting
```
```
Numeric casting is performed on the operands to the inner multiply,
promoting them to the same data type as the accumulator/output.
```

shahidact replied:
Done.

javedabsar1 commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:
```diff
@@ -1065,6 +1060,134 @@ def BatchMatmulOp : LinalgStructuredBase_Op<"batch_matmul", !listconcat([AttrSiz
+      /// Returns a list of AffineMap with the default batch_reduce_matmul indexing charactristic.
+      static SmallVector<AffineMap> getDefaultIndexingMaps(MLIRContext *context);
+
+      /// Returns true if the given broadcast map \p bcastMap is valid for this op.
```
```
/// Returns true if the given `broadcastMap` is valid for this op.
 bool isValidLhsRhsBroadcastMap(AffineMap broadcastMap, bool isLHS = true);
```

javedabsar1 replied:
Is this a legacy API?  How about `isValidLhsRhsMap`?

shahidact replied:
No, we introduced it.

shahidact replied:
We thought of this name to imply that it is validating the broadcast map.

shahidact replied:
Sure, I will update the comment.  I think lhs/rhs is conventionally used for operands.

shahidact replied:
Done.

javedabsar1 commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -5366,6 +5414,169 @@ struct FoldTensorCastUnPackOp : public OpRewritePattern<UnPackOp> {
+         "Expected less than 3 result dim expr.");
+  bool isValid = false;
+  enum Indices { batchPos, mPos, nPos, kPos };
+  if (bcastMap.getNumResults() == 1) {
```
s/exp/expr/ ? otherwise it sounds like 'exponent' on first look.

shahidact replied:
Done.

shahidact committed:
-Updates formatting, comments and variable names.

shahidact commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:
```diff
@@ -1065,6 +1060,134 @@ def BatchMatmulOp : LinalgStructuredBase_Op<"batch_matmul", !listconcat([AttrSiz
+    Numeric casting is performed on the operands to the inner multiply, promoting
+    them to the same data type as the accumulator/output.
+
+    Broadcast and Transpose semantics can be appiled by specifying the explicit attribute
```
Done.

shahidact committed:
-Adds TODO comment to support inference of K dimension
for broadcast map from the existing one.

shahidact commented:
Hi @javedabsar1, @banach-space this PR is ready IMO, pls have a look.

javedabsar1 reviewed (approved):
LGTM. Thanks .

makslevental commented:
congrats @shahidact on this 2 month long journey :smile:

shahidact commented:
> congrats @shahidact on this 2 month long journey 😄

Thanks, Enjoyed it :)
